source/slang/slang-check-expr.cpp

```diff
@@ -3677,4 +3677,43 @@ namespace Slang
         }
     }
 
+    Expr* SemanticsExprVisitor::visitFuncTypeExpr(FuncTypeExpr* expr)
+    {
+        // The input and output to a function type must both be types
+        for(auto& t : expr->parameters)
+            t = CheckProperType(t);
+        expr->result = CheckProperType(expr->result);
+
+        // TODO: Kind checking? Where are we stopping someone passing
+        // constraints around as value-inhabitable types
+
+        // The result of this expression is a `FuncType`, which we need
+        // to wrap in a `TypeType` to indicate that the result is the type
+        // itself and not a value of that type.
+        List<Type*> types;
+        types.reserve(expr->parameters.getCount());
+        for(const auto& t : expr->parameters)
+            types.add(t.type);
+        auto funcType = m_astBuilder->getFuncType(std::move(types), expr->result.type);
+        expr->type = m_astBuilder->getTypeType(funcType);
+
+        return expr;
+    }
+
+    Expr* SemanticsExprVisitor::visitTupleTypeExpr(TupleTypeExpr* expr)
+    {
+        // All tuple members must be types
+        for(auto& t : expr->members)
+            t = CheckProperType(t);
```

Comment:
I think we either need to have `CheckPropertyType` return ErrorType if `t` is null, or handle the case of null here.